test/Pool.test.js

```diff
@@ -1315,6 +1315,196 @@ describe('Pool', function () {
+      assert(err instanceof Promise.TimeoutError);
+      assert(pool.stats().activeTasks = 0);
+    }).always(function () {
+      return pool.exec(add, [1, 2]).then(function () {
+        assert.strictEqual(workerCount, 1);
+      }); 
+    });
+  });
+
+  it('should not terminate worker if abort listener is defined inline worker with Cancellation', function () {
+    var workerCount = 0;
+    var pool = createPool({
+      onCreateWorker: () => {
+        workerCount += 1;
+      },
+      maxWorkers: 1,
+    });
+    function asyncTimeout() {
+      var me = this;
+      return new Promise(function () {
+        let timeout = setTimeout(function() {
+            resolve();
+        }, 5000); 
+        me.worker.addAbortListener(function () {
+          return new Promise(function (resolve) {
+            clearTimeout(timeout);
+            resolve();
+          });
+        });
+      });
+    }
+    function add(a, b) { }
+    const task = pool.exec(asyncTimeout, [],  {
+    })
+    return new Promise(function(resolve) {
+      setTimeout(function() {
+        resolve();
+      }, 50);
+    }).then(function() {
+      return task
+      .cancel()
+      .catch(function(err) {
+        assert(err instanceof Promise.TimeoutError);
+        assert(pool.stats().activeTasks = 0);
+      }).always(function () {
+        return pool.exec(add, [1, 2]).then(function () {
+          assert.strictEqual(workerCount, 1);
+        }); 
+      });
+    });
+
+  });
+
+  it('should invoke timeout for abort handler if timeout period is reached with Timeout', function () {
+    var workerCount = 0;
+    var pool = createPool(__dirname + '/workers/cleanup-abort.js', {
+      maxWorkers: 1,
+      onCreateWorker: function() {
+        workerCount += 1;
+      }
+    });
+  
+    return pool.exec('asyncAbortHandlerNeverResolves', [])
+    .timeout(1000)
+    .catch(function (err) {
+      assert(err instanceof Promise.TimeoutError);
+
+      var stats = pool.stats();
+      assert(stats.busyWorkers === 1);
+    }).always(function() {
+      return pool.exec(add, [1, 2]).then(function() {
+        assert.strictEqual(workerCount, 1);
+      });
+    });
+  });
+
+
+  it('should invoke timeout for abort handler if timeout period is reached with Cancellation', function () {
+    var workerCount = 0;
+    var pool = createPool(__dirname + '/workers/cleanup-abort.js', {
+      maxWorkers: 1,
+      onCreateWorker: function() {
+        workerCount += 1;
+      }
+    });
+  
+    const task = pool.exec('asyncAbortHandlerNeverResolves', [])
+    
+    return new Promise(function(resolve) {
+      setTimeout(function() {
+        resolve();
+      }, 50);
+    }).then(function() { 
+      return task.cancel()
+      .catch(function (err) {
+        assert(err instanceof Promise.TimeoutError);
+        var stats = pool.stats();
+        assert(stats.busyWorkers === 1);
+      }).always(function() {
+        return pool.exec(add, [1, 2]).then(function() {
```

Comment:
I'm really a bit confused about all the `pool.exec(add, [1, 2])` in the unit tests, I would love to see a more explicit solution like with `stats().totalWorkers`. Or maybe I'm totally misunderstanding the reason for `pool.exec(add, [1, 2])` in the unit tests 😅 .

Reply:
I feel using calling another such as `add` is required to fully test this new feature as the tests are trying to prove that that a new worker is never created after the first, and due to the design of `pool` we create new workers at time of need. So by calling another method, such as `add` we can assert on this behavior and show that the pool resource provisioning is working as expected. We can still add calls to `pool.stat()` and assert on the worker counts. But I feel that the invocation of another method is beneficial to assert the pool is managing `worker` instances in the ways we expect.

Reply:
That is true, it is a good extra test indeed calling a method right after it. I think we want to ensure two things happen:

1. After abort, a worker is _not_ terminated, so workerCount remains 1 and should not become 0.
2. When executing another task (like `add`) after aborting, the existing worker must be used. I think for this it makes sense to set `maxWorkers: 2` (so there is room for workerpool to mistakenly create a 2nd worker) and ensure the workerCount is still 1 after executing `add`.

I think by only testing `workerCount === 1` _after_ executing `add`, we will miss when aborting would not work correctly and would terminate the worker. I think it is essential to verify the worker count before and after `add`.

Reply:
Good points!